File: src/16-challenge/solution.py
def fn(gifts: list[int], maxWeight: int) -> int | None:
    if not gifts:
        return 0

    sleighs_remaining_weight: list[int] = [maxWeight]

    for gift_weight in gifts:
        if gift_weight > maxWeight:
            return None

        remaining_weight = sleighs_remaining_weight[-1] - gift_weight
        if remaining_weight >= 0:
            sleighs_remaining_weight[-1] = remaining_weight
            continue

        sleighs_remaining_weight.append(maxWeight - gift_weight)

    return len(sleighs_remaining_weight)

File: src/16-challenge/solution.ts
type Gifts = number[]
type MaxWeight = number
type Result = number | null

export default function fn(gifts: Gifts, maxWeight: MaxWeight): Result {
	if (!gifts.length) return 0

	const sleighsRemainingWeight: number[] = [maxWeight]

	for (const giftWeight of gifts) {
		if (giftWeight > maxWeight) return null

		const remainingWeight = sleighsRemainingWeight.at(-1)! - giftWeight
		if (remainingWeight >= 0) {
			sleighsRemainingWeight[sleighsRemainingWeight.length - 1] = remainingWeight
			continue
		}

		sleighsRemainingWeight.push(maxWeight - giftWeight)
	}

	return sleighsRemainingWeight.length
}
